### DISCUSSION TITLE
2 주차 사전 과제, 김경진

### DISCUSSION THREAD

**Discussion by lessmadmax** [MrMirror21/2024-KHU-CSE-TrackStudy-Web#23] [General]
**일상생활에서 “상태(state)” 라는 개념이 사용되는 예시를 1가지 작성해주세요. 각 예시에서 “상태” 가 어떻게 변하고, 그 변화가 어떤 결과를 가져오는지 적어주시면 됩니다.**

일상생활에서 상태라는 개념을 사용하는 예시로 날씨를 들 수 있습니다. 날씨는 다양한 상태로 변화할 수 있는데, 흔히 표현하는 바로는 좋다, 나쁘다가 있습니다. 날씨가 좋아지면 사람들의 야외 활동이 많아집니다. 날씨가 나빠지면 사람들이 실내에서 주로 시간을 보내게 됩니다.

**자신이 주로 사용하는 웹 페이지의 이름과, 그 웹 페이지에서 사용자의 입력에 따라 즉시 반응하는 요소들의 예를 3가지 작성해주세요. 각 요소가 어떤 사용자 행동에 반응하고, 어떻게 변화하는지 설명해주세요.**

저는 유튜브를 자주 사용합니다. 사용자의 입력에 따라 즉시 반응하는 요소들은 아래와 같습니다.
1. 검색 : 제가 원하는 검색어를 입력하면 페이지가 리로딩되면서 해당되는 영상들을 페이지에 띄워줍니다.
2. 알림 표시 버튼 : 버튼을 클릭하면 페이지를 재구성하지 않고 알림을 확인하는 창을 띄워줍니다.
3. 좋아요 버튼 : 버튼을 클릭하면 그에 반응하여 영상의 좋아요 개수가 늘고 다시 클릭하면 좋아요를 취소합니다.

**JavaScript로 작성된 다음 코드를 실행했을 때 console.log(x.count) 의 결과를 작성하고, 그 결과가 나온 이유를 작성해주세요.**
```
let x = { count: 0 };
let y = x;
y.count = 1;
console.log(x.count);
```
출력 결과는 1입니다. x를 y에 할당하게 되면 x와 y는 똑같은 객체를 가리키게 됩니다. 따라서 y.count를 1로 변경하게 되면 x.count를 1로 변경한 것과 똑같은 결과를 낳게 됩니다. 

**2주차 스터디 도중 props의 특성 중 “불변성(immutability)”에 대해서 설명드렸습니다. 프로그래밍에서 불변성을 지키는 것이 중요할까요? 자신의 의견과 이유를 적어주세요.**

중요하다고 생각합니다. 프로그램의 요소들이 쉽게 변할 경우, 프로그램의 동작이 예측하기 힘들어지면서 예기치 않은 문제발생의 가능성이 있기 때문입니다.

**JavaScript로 작성된 다음 코드를 실행했을 때 어떤 순서로 콘솔에 출력될 지 예측해보세요. 자신이 예측했던 결과와 그 이유, 그리고 실제 출력된 결과를 작성해주세요.**
```
console.log('시작');

setTimeout(() => {
  console.log('3초 후');
}, 3000);

setTimeout(() => {
  console.log('1초 후');
}, 1000);

console.log('끝');
```
예상 출력 결과 : 
시작
끝
1초 후
3초 후

출력 결과 : 
시작 
끝
1초 후
3초 후

setTimeout 함수는 지연 시간 후에 실행될 함수를 첫번째 인수로 받고 첫번째 인수로 받은 함수를 실행시킬 때까지 기다릴 시간을 두번째 인수로 받습니다. 해당 시간은 밀리초 단위입니다. 따라서 코드 입력 즉시 시작과 끝이 출력되며 1000밀리초가 지난 뒤 1초 후가 출력되고 3000밀리초가 지난 뒤 3초 후가 출력됩니다.

**Comment by MrMirror21** [comment]
작성해주신 사전과제 잘 봤습니다. 전체적으로 잘 답변해주셨고, shallow copy, deep copy 문제나 JavaScript의 비동기 문제도 잘 풀어주신 것 같습니다. 지난 3주차 스터디 내용을 이해하는데 도움이 되셨으면 좋겠네요. 남은 스터디도 많은 참여 부탁드립니다👍
